File: external/boost_1.55.0/libs/python/test/pickle4.cpp
// Copyright Ralf W. Grosse-Kunstleve 2004. Distributed under the Boost
// Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

/*
    This example shows how to enable pickling without using the
    pickle_suite. The pickling interface (__getinitargs__) is
    implemented in Python.

    For more information refer to boost/libs/python/doc/pickle.html.
 */

#include <boost/python/module.hpp>
#include <boost/python/class.hpp>

#include <string>

namespace boost_python_test {

  // A friendly class.
  class world
  {
    private:
      std::string country;
    public:
      world(const std::string& _country) {
        this->country = _country;
      }
      std::string greet() const { return "Hello from " + country + "!"; }
      std::string get_country() const { return country; }
  };

}

BOOST_PYTHON_MODULE(pickle4_ext)
{
  using namespace boost::python;
  using namespace boost_python_test;
  class_<world>("world", init<const std::string&>())
      .enable_pickling()
      .def("greet", &world::greet)
      .def("get_country", &world::get_country)
      ;
}

File: external/boost_1.55.0/libs/python/test/pickle4.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
r'''>>> import pickle4_ext
    >>> import pickle
    >>> def world_getinitargs(self):
    ...   return (self.get_country(),)
    >>> pickle4_ext.world.__getinitargs__ = world_getinitargs
    >>> pickle4_ext.world.__module__
    'pickle4_ext'
    >>> pickle4_ext.world.__safe_for_unpickling__
    1
    >>> pickle4_ext.world.__name__
    'world'
    >>> pickle4_ext.world('Hello').__reduce__()
    (<class 'pickle4_ext.world'>, ('Hello',))
    >>> wd = pickle4_ext.world('California')
    >>> pstr = pickle.dumps(wd)
    >>> wl = pickle.loads(pstr)
    >>> print wd.greet()
    Hello from California!
    >>> print wl.greet()
    Hello from California!
'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)
